esi_leap/objects/contract.py

```diff
@@ -48,9 +52,41 @@ def get_all(cls, context, filters):
 
     def create(self, context=None):
         updates = self.obj_get_changes()
+        updates['uuid'] = uuidutils.generate_uuid()
+
+        if 'offer_uuid' not in updates:
+            raise exception.ContractNoOfferUUID()
+
+        related_offer = self.dbapi.offer_get(updates['offer_uuid'])
+        if related_offer is None:
+            raise exception.OfferNotFound(offer_uuid=updates['offer_uuid'])
+
+        if related_offer.status != statuses.AVAILABLE:
+            raise exception.OfferNotAvailable(offer_uuid=related_offer.uuid,
+                                              status=related_offer.status)
+
+        if 'start_time' not in updates:
+            updates['start_time'] = datetime.datetime.now()
+
+        if 'end_time' not in updates:
+            q = self.dbapi.offer_get_first_availability(related_offer,
+                                                        updates['start_time'])
+            if q is None:
+                updates['end_time'] = related_offer.end_time
+            else:
+                updates['end_time'] = q.start_time
+
+        self.dbapi.offer_verify_availability(related_offer,
+                                             updates['start_time'],
+                                             updates['end_time'])
+
         db_contract = self.dbapi.contract_create(updates)
         self._from_db_object(context, self, db_contract)
 
+    def cancel(self):
+        self.status = statuses.CANCELLED
+        self.save(None)
```

Comment:
I think you have to go through some of the contract's `expire` logic; if the contract is currently - active? - then you'll want to unassign the resource, etc etc.

Reply:
This has been addressed